dvc/repo/reproduce.py

```diff
@@ -191,40 +200,10 @@ def _reproduce_stages(  # noqa: C901
     return result
 
 
-def _get_steps(graph, stages, downstream, single_item):
-    import networkx as nx
-
-    active = graph.copy()
-    if not single_item:
-        # NOTE: frozen stages don't matter for single_item
-        for stage in graph:
-            if stage.frozen:
-                # NOTE: disconnect frozen stage from its dependencies
-                active.remove_edges_from(graph.out_edges(stage))
-
-    all_pipelines: List["Stage"] = []
-    for stage in stages:
-        if downstream:
-            # NOTE (py3 only):
-            # Python's `deepcopy` defaults to pickle/unpickle the object.
-            # Stages are complex objects (with references to `repo`,
-            # `outs`, and `deps`) that cause struggles when you try
-            # to serialize them. We need to create a copy of the graph
-            # itself, and then reverse it, instead of using
-            # graph.reverse() directly because it calls `deepcopy`
-            # underneath -- unless copy=False is specified.
-            nodes = nx.dfs_postorder_nodes(active.reverse(copy=False), stage)
-            all_pipelines += reversed(list(nodes))
-        else:
-            all_pipelines += nx.dfs_postorder_nodes(active, stage)
-
-    steps = []
-    for stage in all_pipelines:
-        if stage not in steps:
-            # NOTE: order of steps still matters for single_item
-            if single_item and stage not in stages:
-                continue
-
-            steps.append(stage)
-
-    return steps
+def _remove_frozen_stages(graph: "DiGraph") -> "DiGraph":
```

Comment:
Minor: Maybe `frozen` could be an option of `get_steps`? Would be easier to test

Reply:
I want `get_steps()` to be generic. `frozen` is tested through functional tests.

Reply:
> I want `get_steps()` to be generic. `frozen` is tested through functional tests.

I got it, no strong opinion. 
It is just that it already includes `downstream` so I don't see the difference that makes `frozen` less generic than `downstream`.

And easier I also meant "simpler" as in not needing functional to also check for how `frozen` affects the graph (as you do for `downstream` in the new unit tests

Reply:
Calculating downstream node is a generic graph operation, so it is not quite the same as `--downstream` flag.

Reply:
Whereas, `frozen` is a property of a `Stage`. Bootstrapping those are harder, and also harder to test. I'd just leave it to the functional tests that we have.